lib/xgu/xgu.h

```diff
@@ -804,71 +826,81 @@ uint32_t* xgu_set_texture_format(uint32_t* p, uint8_t context_dma, bool cubemap_
+                                 bool r_signed, bool g_signed, bool b_signed, bool a_signed) {
+    assert(texture_index < 4);
+    return push_command_parameter(p, NV097_SET_TEXTURE_FILTER + texture_index*64,
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_MIPMAP_LOD_BIAS, lod_bias) |
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_MIN, filter_min) |
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_MAG, filter_mag) |
-                                  XGU_MASK(NV097_SET_TEXTURE_FILTER_ASIGNED, alpha_signed) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FILTER_ASIGNED, a_signed) |
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_RSIGNED, r_signed) |
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_GSIGNED, g_signed) |
                                   XGU_MASK(NV097_SET_TEXTURE_FILTER_BSIGNED, b_signed));
 }
 
 inline
-uint32_t* xgu_set_texture_image_rect(uint32_t* p, uint16_t width, uint16_t height) {
-    return push_command_parameter(p, NV097_SET_TEXTURE_IMAGE_RECT,
+uint32_t* xgu_set_texture_image_rect(uint32_t* p, unsigned int texture_index, uint16_t width, uint16_t height) {
+    assert(texture_index < 4);
+    return push_command_parameter(p, NV097_SET_TEXTURE_IMAGE_RECT + texture_index*64,
                                   XGU_MASK(NV097_SET_TEXTURE_IMAGE_RECT_WIDTH, width) |
                                   XGU_MASK(NV097_SET_TEXTURE_IMAGE_RECT_HEIGHT, height));
 }
 
 inline
-uint32_t* xgu_set_texture_palette(uint32_t* p, bool context_dma, XguPaletteLen palette_length,
-                                  uint32_t offset) {
-    return push_command_parameter(p, NV097_SET_TEXTURE_palette,
+uint32_t* xgu_set_texture_palette(uint32_t* p, unsigned int texture_index, bool context_dma, XguPaletteLen palette_length,
+                                  const void* offset) {
+    assert(texture_index < 4);
+    return push_command_parameter(p, NV097_SET_TEXTURE_PALETTE + texture_index*64,
                                   XGU_MASK(NV097_SET_TEXTURE_PALETTE_CONTEXT_DMA, context_dma) |
                                   XGU_MASK(NV097_SET_TEXTURE_PALETTE_LENGTH, palette_length) |
-                                  XGU_MASK(NV097_SET_TEXTURE_PALETTE_OFFSET, offset));
+                                  XGU_MASK(NV097_SET_TEXTURE_PALETTE_OFFSET, (uint32_t)offset)); //FIXME: Probably bad offset? I don't think this has to be shifted?
 }
 
 inline
-uint32_t* xgu_set_texture_border_color(uint32_t* p, uint32_t color) {
-    return push_command_parameter(p, NV097_SET_TEXTURE_BORDER_COLOR, color);
+uint32_t* xgu_set_texture_border_color(uint32_t* p, unsigned int texture_index, uint32_t color) {
+    assert(texture_index < 4);
+    return push_command_parameter(p, NV097_SET_TEXTURE_BORDER_COLOR + texture_index*64, color);
```

Comment:
Other color values are now described with three `float`s. Should that be the case here too?

Reply:
no, because they are always described in the GPU format.
The other instances are 3 floats - this one isn't (it's probably 4 uint8_t, but maybe also something else, so I'm not changing it until we test it).